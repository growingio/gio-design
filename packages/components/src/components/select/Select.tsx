import React, { useState, useContext, useMemo, useRef, useEffect, useCallback } from 'react';
import classnames from 'classnames';

import { DownFilled, BoxFilled } from '@gio-design/icons';
import { filter, isNil, without, uniqueId, uniqBy, findIndex } from 'lodash';
import { SizeContext } from '../config-provider/SizeContext';
import Dropdown from '../dropdown';
import Tag from '../tag';
import List from '../list';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';

import { SelectProps, Option, MaybeArray } from './interface';

const defaultArrowComponent = <DownFilled />;

const defaultListRowHeight = 44;

const customOptionKeyPrefix = 'select_custom_option_';
const customOptionKey = uniqueId(customOptionKeyPrefix);

export const CustomOption = (value: string, withGroup = false, id = customOptionKeyPrefix): Option =>
  withGroup
    ? {
        value,
        label: value,
        groupValue: id,
        groupLabel: '自由输入',
      }
    : {
        value,
        label: value,
      };

// provide search matching hightlight;
export const defaultLabelRenderer = (input: string, prefix: string) => (
  option: Option,
  isGroup: boolean
): React.ReactNode => {
  if (isGroup) return option.label;
  const index = option.label.indexOf(input);
  return (
    <div>
      {option.label.slice(0, index)}
      <span className={`${prefix}-search-highlight`}>{option.label.slice(index, index + input.length)}</span>
      {option.label.slice(index + input.length)}
    </div>
  );
};

const defaultSearchPredicate = (input: string) => (o: Option) => o.label.includes(input);

const defaultMatchPredicate = (input: string) => (o: Option) => o.label === input;

const defaultNotFoundContent = (
  <div
    style={{
      width: '100%',
      textAlign: 'center',
      color: '#a3adc8',
      padding: '68px 0',
    }}
  >
    <BoxFilled size="48px" style={{ marginBottom: 24 }} />
    <div style={{ fontSize: 12 }}>暂无选项...</div>
  </div>
);

const defaultOptionLabelRenderer = (value: string, option?: Option) => option?.label || value;

const Select: React.FC<SelectProps> = (props: SelectProps) => {
  const sizeContext = useContext(SizeContext);
  const {
    size = sizeContext || 'middle',
    options = [],
    multiple = false,
    placeholder,
    searchable = false,
    disabled = false,
    allowCustomOption = false,
    notFoundContent = defaultNotFoundContent,
    customizePrefixCls,
    className,
    style,
    bordered = true,
    arrowComponent = defaultArrowComponent,
    autoWidth = true,
    listHeight,
    listRowHeight = defaultListRowHeight,
    labelRenderer = defaultLabelRenderer,
    searchPredicate = defaultSearchPredicate,
    matchPredicate = defaultMatchPredicate,
    optionLabelRenderer = defaultOptionLabelRenderer,
    defaultValue,
    value: controlledValue,
    onChange,
    onSearch,
    onSelect,
    onDeselect,
    getContainer,
    dropDownVisible,
    onDropDownVisibleChange,
    dropDownClassName,
    dropDownStyle,
  } = props;

  const prefix = usePrefixCls('select', customizePrefixCls);

  const [unControlledValue, setUnControlledValue] = useState(defaultValue);
  const isControlled = !isNil(controlledValue);
  const value = isControlled ? controlledValue : unControlledValue;
  const [isFocused, setFocused] = useState(false);
  const [_visible, _setVisible] = useState(false);
  const visible = isNil(dropDownVisible) ? _visible : dropDownVisible;
  const setVisbile = isNil(onDropDownVisibleChange) ? _setVisible : onDropDownVisibleChange;
  const [input, setInput] = useState('');
  const inputRef = useRef<HTMLInputElement>(null);
  const [inputWidth, setInputWidth] = useState(2);
  const inputWidthRef = useRef<HTMLDivElement>(null);
  const selectorRef = useRef<HTMLDivElement>(null);

  const clearInput = () => {
    setInput('');
  };

  const onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();
    onSearch?.(e.target.value);
    setInput(e.target.value);
  };

  const onInputClick = (e: React.MouseEvent<HTMLInputElement>) => {
    if (input) e.stopPropagation();
  };

  const [valueToOptionMap, hasGroup] = useMemo(() => {
    let group = false;
    const map = options.reduce((m, option) => {
      if (option.groupLabel) group = true;
      m.set(option.value, option);
      return m;
    }, new Map());
    return [map, group];
  }, [options]);

  useEffect(() => {
    if (!disabled) {
      setFocused(visible);
      if (visible) {
        if (inputRef.current) {
          inputRef.current.focus();
        }
      }
    }
  }, [visible, disabled]);

  const onVisibleChange = (optVisible: boolean) => {
    if (!disabled) {
      setVisbile(optVisible);
      if (!optVisible) {
        setTimeout(clearInput, 0);
        inputRef.current?.blur();
      }
    }
  };

  const getOptionByValue = useCallback(
    (optValue: string): Option | undefined => {
      return valueToOptionMap.get(optValue);
    },
    [valueToOptionMap]
  );

  const getOptionsByValue = (optValue: MaybeArray<string>): MaybeArray<Option> | undefined => {
    return Array.isArray(optValue)
      ? optValue.reduce((prev: Option[], v) => {
          const op = getOptionByValue(v);
          if (op) {
            prev.push(op);
          }
          return prev;
        }, [])
      : getOptionByValue(optValue);
  };

  useEffect(() => {
    if (inputWidthRef.current) {
      setInputWidth(inputWidthRef.current?.getBoundingClientRect().width + 4);
    }
  }, [input]);

  const extendedOptions = useMemo(() => {
    const result: Option[] = [];
    if (Array.isArray(value) && allowCustomOption) {
      value.forEach((v) => {
        const op = getOptionByValue(v);
        if (!op) {
          result.push(CustomOption(v, hasGroup, customOptionKey));
        }
      });
    }
    return [...options, ...result];
  }, [options, value, getOptionByValue, allowCustomOption, hasGroup]);

  const filteredOptions = useMemo(() => filter(extendedOptions, searchPredicate(input)), [
    searchPredicate,
    extendedOptions,
    input,
  ]);

  const hasExactMatch = useMemo(() => findIndex(filteredOptions, matchPredicate(input)) > -1, [
    matchPredicate,
    filteredOptions,
    input,
  ]);

  const completeOptions = useMemo(
    () =>
      !!input && !hasExactMatch && allowCustomOption
        ? [CustomOption(input, hasGroup, customOptionKey), ...filteredOptions]
        : filteredOptions,
    [hasExactMatch, allowCustomOption, filteredOptions, input, hasGroup]
  );

  const groupCount = useMemo(() => (hasGroup ? uniqBy(completeOptions, 'groupValue').length : 0), [
    completeOptions,
    hasGroup,
  ]);

  const onValueChange = (optValue: MaybeArray<string>) => {
    if (!isControlled) {
      setUnControlledValue(optValue);
    }
    onChange?.(optValue, getOptionsByValue(optValue));
  };

  const onSelectionKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    switch (e.keyCode) {
      // delete key
      case 8:
      case 46:
        if (!input.length && value) {
          if (Array.isArray(value)) {
            value.pop();
            onValueChange([...value]);
          } else {
            onValueChange('');
          }
        }
        break;
      // enter key
      case 13:
        if (!hasExactMatch && allowCustomOption && input) {
          if (Array.isArray(value)) {
            value.push(input);
            onValueChange([...(value as Array<string>)]);
          }
        }
        clearInput();
        break;
      default:
        break;
    }
  };

  const onListSelect = (selectedValue: string, _: string[], option: Option) => {
    if (!multiple) {
      onVisibleChange(false);
    }
    if (multiple) {
      inputRef.current?.focus();
    } else {
      inputRef.current?.blur();
    }
    clearInput();
    onSelect?.(selectedValue, option);
  };

  const onListDeselect = (selectedValue: string, _: string[], option: Option) => {
    if (!multiple) {
      onVisibleChange(false);
    }
    if (multiple) {
      inputRef.current?.focus();
    } else {
      inputRef.current?.blur();
    }
    onDeselect?.(selectedValue, option);
  };

  const onTagCloseClick = (e: React.MouseEvent<HTMLSpanElement, MouseEvent>, v: string) => {
    e.stopPropagation();
    onValueChange(without(value || [], v));
  };

  const renderSingleValue = () =>
    !input && typeof value === 'string' ? (
      <div className={`${prefix}-item`}>
        <span className={`${prefix}-item-text`}>{optionLabelRenderer(value, getOptionByValue(value))}</span>
      </div>
    ) : null;

  const renderMultipleValue = () =>
    (value as Array<string>)?.map((v) => (
      <Tag
        key={v}
        className={`${prefix}-item`}
        persistCloseIcon
        closable
        onClose={(e) => onTagCloseClick(e, v)}
        disabled={disabled}
      >
        {optionLabelRenderer(v, getOptionByValue(v))}
      </Tag>
    ));

  const renderSearchInput = () => (
    <>
      <div ref={inputWidthRef} className={classnames(`${prefix}-input-reference`, `${prefix}-item`)}>
        {input}
      </div>
      <input
        ref={inputRef}
        style={{ width: inputWidth }}
        className={classnames(`${prefix}-input`, `${prefix}-item`)}
        value={input}
        onChange={onInputChange}
        tabIndex={-1}
        onKeyDown={onSelectionKeyDown}
        onClick={onInputClick}
      />
    </>
  );

  const renderPlaceHolder = () => {
    return (!value || !value.length) && !input && placeholder ? (
      <div className={`${prefix}-item ${prefix}-placeholder`}>{placeholder}</div>
    ) : null;
  };

  const trigger = (
    <div
      role={disabled ? undefined : 'combobox'}
      aria-expanded={visible}
      aria-controls="expandable"
      className={classnames(`${prefix}`, `${prefix}-${size}`, {
        [`${prefix}-single`]: !multiple,
        [`${prefix}-focused`]: isFocused,
        [`${prefix}-disabled`]: disabled,
        [`${prefix}-bordered`]: bordered,
        className,
      })}
      aria-disabled={disabled}
      style={style}
      ref={selectorRef}
    >
      <div className={`${prefix}-selector`}>
        <div className={classnames(`${prefix}-values-wrapper`)}>
          {multiple ? renderMultipleValue() : renderSingleValue()}
          {searchable && !disabled && renderSearchInput()}
          {renderPlaceHolder()}
        </div>
      </div>
      <div className={`${prefix}-arrow`}>{arrowComponent}</div>
    </div>
  );

  const list = (
    <div style={{ width: autoWidth ? Math.max(selectorRef.current?.clientWidth || 0, 160) : undefined }}>
      {completeOptions.length > 0 ? (
        <List
          value={value}
          width={autoWidth ? Math.max(selectorRef.current?.clientWidth || 0, 160) : undefined}
          dataSource={completeOptions}
          onChange={onValueChange}
          isMultiple={multiple}
          labelRenderer={labelRenderer(input, prefix)}
          onSelect={onListSelect}
          onDeselect={onListDeselect}
          height={listHeight || (completeOptions.length + groupCount) * listRowHeight}
          rowHeight={listRowHeight}
        />
      ) : (
        notFoundContent
      )}
    </div>
  );

  return (
    <Dropdown
      visible={visible}
      onVisibleChange={onVisibleChange}
      trigger={['click']}
      placement="bottomLeft"
      overlay={list}
      overlayClassName={classnames(`${prefix}-dropdown`, dropDownClassName)}
      overlayInnerStyle={dropDownStyle}
      getTooltipContainer={getContainer}
    >
      {trigger}
    </Dropdown>
  );
};

export default Select;
