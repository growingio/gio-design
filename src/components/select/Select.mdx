import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Select from './Select';

# Select 选择器

<Subtitle>
  下拉选择器，让用户做出选择。Select 与 Dropdown 的区别是 Select 承载的是选项，Dropdown 的承载的是操作。Select
  有三种尺寸的高度（大：40px、中：36px、小：30px），同 button。
</Subtitle>

## 参数说明

<ArgsTable of={Select} />

## 代码演示

### 默认

根据使用场景选用紧凑 padding 型的 Select，或者 固定宽度的 Select。Select 的下拉菜单最小宽度为 160px。

[Example](/?path=/story/functional-components-select--default)

### 多选

1. 可以通过输入及下拉选项卡进行多选。多选项以 Tag 的形式出现在输入框内。每个选项 Tag 都可以通过「X」删除。
2. Select 输入框的宽度根据页面需要设定，Select 的下拉菜单最小宽度为 160px。

[Example](/?path=/story/functional-components-select--multiple)

### 无边框背景的 Select

当页面的整洁度要求很高时，可以选用没有边框背景的 Select。

[Example](/?path=/story/functional-components-select--no-border)
