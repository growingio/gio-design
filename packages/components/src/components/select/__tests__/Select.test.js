import React from 'react';
import '../../../../es/components/select/style/index.css';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Select from '../index';
import { CustomOption, defaultLabelRenderer } from '../Select';

const labels = ['全部', '已上线', '待上线', '已下线', '草稿'];
const values = ['all', 'online', 'pending', 'off', 'draft'];

const options = values.map((value, index) => ({
  value,
  label: labels[index],
  groupValue: 'platform',
  groupLabel: '应用平台',
}));

const optionsWithOutGroup = values.map((value, index) => ({
  value,
  label: labels[index],
}));

describe('<Select />', () => {
  // For List use AutoSizer.
  // AutoSizer uses offsetWidth and offsetHeight.
  // Jest runs in JSDom which doesn't support measurements APIs.
  beforeEach(() => {
    Object.defineProperty(HTMLElement.prototype, 'offsetHeight', { configurable: true, value: 320 });
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', { configurable: true, value: 160 });
  });

  it('renders <Select /> components', () => {
    const tree = renderer.create(<Select defaultValue="all" options={options} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders <Select /> components', () => {
    const tree = renderer.create(<Select defaultValue="all" options={optionsWithOutGroup} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have correc dom structure', () => {
    const tree = shallow(<Select options={options} />);
    expect(tree.find('.gio-select').exists()).toBeTruthy();
    expect(tree.find('.gio-select-selector').exists()).toBeTruthy();
    expect(tree.find('.gio-select-values-wrapper').exists()).toBeTruthy();
    expect(tree.find('.gio-select-arrow').exists()).toBeTruthy();
  });

  it('should have correc dom structure', () => {
    const tree = shallow(<Select options={optionsWithOutGroup} />);
    expect(tree.find('.gio-select').exists()).toBeTruthy();
    expect(tree.find('.gio-select-selector').exists()).toBeTruthy();
    expect(tree.find('.gio-select-values-wrapper').exists()).toBeTruthy();
    expect(tree.find('.gio-select-arrow').exists()).toBeTruthy();
    tree.unmount();
  });

  it('should have correct classes', () => {
    const tree = shallow(<Select options={options} />);
    tree.setProps({ size: 'small' });
    expect(tree.find('.gio-select-small').exists()).toBeTruthy();
    tree.setProps({ size: 'middle' });
    expect(tree.find('.gio-select-middle').exists()).toBeTruthy();
    tree.setProps({ size: 'large' });
    expect(tree.find('.gio-select-large').exists()).toBeTruthy();
  });

  it('select dropdown should display correct search result', () => {
    const tree = mount(<Select options={options} searchable allowCustomOption />);
    expect(document.body.querySelector('.gio-select-dropdown')).toBeNull();
    act(() => {
      tree.simulate('click');
    });
    const dropdown = document.querySelector('.gio-select-dropdown');
    expect(dropdown).not.toBeNull();
    act(() => {
      tree.find('input').simulate('change', { target: { value: '全' } });
    });
    act(() => {
      tree.find('input').simulate('click');
    });
    expect(dropdown.querySelectorAll('.gio-select-option')).toHaveLength(4);
    tree.unmount();
  });
});

describe('<Select multiple/>', () => {
  it('renders <Select multiple/> components', () => {
    const tree = renderer.create(<Select defaultValue={['all']} options={options} multiple />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders <Select multiple/> components', () => {
    const tree = renderer.create(<Select defaultValue={['all']} options={optionsWithOutGroup} multiple />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('select dropdown should display correct search result', () => {
    const tree = mount(<Select options={options} searchable />);
    act(() => {
      tree.simulate('click');
    });
    const dropdown = document.querySelector('.gio-select-dropdown');
    expect(dropdown).not.toBeNull();
    act(() => {
      tree.find('input').simulate('change', { target: { value: '全' } });
    });
    expect(document.querySelectorAll('.gio-select-dropdown .gio-select-option')).toHaveLength(2);
    act(() => {
      tree.find('.gio-select-dropdown .gio-select-option').at(0).simulate('click');
    });
    act(() => {
      expect(tree.find('.gio-select-input-reference.gio-select-item').text()).toBe('全');
    });
    act(() => {
      tree.unmount();
    });
  });
});

describe('<Select /> callback functions should work as expected', () => {
  const onChange = (v, o) => {
    expect(v).toBe('all');
    expect(o).toMatchObject({ label: '全部', value: 'all' });
  };

  const onSelect = (v, o) => {
    expect(v).toBe('all');
    expect(o).toMatchObject({ label: '全部', value: 'all' });
  };

  const onDeSelect = (v, o) => {
    expect(v).toBe('all');
    expect(o).toMatchObject({ label: '全部', value: 'all' });
  };

  const onSearch = (i) => {
    expect(i).toBe('全部');
  };

  it('onChange, onSelect, onDeselect, OnSearch', () => {
    const tree = mount(
      <Select
        options={optionsWithOutGroup}
        onChange={onChange}
        onSelect={onSelect}
        onDeselect={onDeSelect}
        onSearch={onSearch}
        searchable
        allowCustomOption
      />
    );
    act(() => {
      tree.simulate('click');
    });
    act(() => {
      tree.find('input').simulate('change', { target: { value: '全部' } });
    });
    act(() => {
      document.querySelector('.gio-select-dropdown .gio-select-option').click();
    });
    tree.unmount();
  });
});

describe('<Select Multiple /> callback functions should work as expected on multiple mode', () => {
  const onChange = (v, o) => {
    expect(v).toEqual(['online', 'all']);
    expect(o).toEqual([
      { value: 'online', label: '已上线' },
      { value: 'all', label: '全部' },
    ]);
  };

  const onSelect = (v, o) => {
    expect(v).toBe('all');
    expect(o).toMatchObject({ label: '全部', value: 'all' });
  };

  const onDeSelect = (v, o) => {
    expect(v).toBe('all');
    expect(o).toMatchObject({ label: '全部', value: 'all' });
  };

  const onSearch = (i) => {
    expect(i).toBe('全部');
  };

  it('onChange, onSelect, onDeselect, OnSearch', () => {
    const tree = mount(
      <Select
        multiple
        options={optionsWithOutGroup}
        onChange={onChange}
        onSelect={onSelect}
        onDeselect={onDeSelect}
        onSearch={onSearch}
        searchable
        defaultValue={['online']}
      />
    );
    act(() => {
      tree.simulate('click');
    });
    act(() => {
      tree.find('input').simulate('change', { target: { value: '全部' } });
    });
    act(() => {
      document.querySelector('.gio-select-dropdown .gio-select-option').click();
    });
    tree.unmount();
  });
});

describe('<Select allowCustomOptions multiple/> can create option by presee enter, and unselect by press delete key', () => {
  it('should be able to create by enter', () => {
    const tree = mount(<Select multiple allowCustomOption searchable options={options} defaultValue={['all']} />);
    act(() => {
      tree.find('input').simulate('change', { target: { value: 'test' } });
    });
    act(() => {
      tree.find('input').simulate('keydown', { keyCode: 13 });
    });
    expect(tree.render().find('.gio-select-values-wrapper').children('.gio-tag')).toHaveLength(2);
    act(() => {
      tree.find('input').simulate('change', { target: { value: '' } });
    });
    act(() => {
      tree.find('input').simulate('keydown', { keyCode: 46 });
    });
    expect(tree.render().find('.gio-select-values-wrapper').children('.gio-tag')).toHaveLength(1);
    act(() => {
      tree.unmount();
    });
  });
});

describe('<Select /> when press delete key will unselect current option', () => {
  it('should be able to create by enter', () => {
    const tree = mount(<Select searchable options={options} defaultValue="all" />);
    act(() => {
      tree.find('input').simulate('keydown', { keyCode: 46 });
    });
    expect(tree.render().find('.gio-select-values-wrapper').children('.gio-select-item-text').text()).toBe('');
    act(() => {
      tree.unmount();
    });
  });
});

describe('<Select /> when press delete key will unselect current option', () => {
  it('should be able to create by enter', () => {
    const onDeSelect = (v, o) => {
      expect(v).toBe('all');
      expect(o).toMatchObject({ label: '全部', value: 'all' });
    };

    const tree = mount(<Select multiple searchable options={options} defaultValue={['all']} onDeselect={onDeSelect} />);

    act(() => {
      tree.find('.gio-tag .gio-tag-closable-icon').at(0).simulate('click');
    });

    expect(tree.render().find('.gio-select-values-wrapper').children()).toHaveLength(2);
    act(() => {
      tree.unmount();
    });
  });
});

describe('<Select /> deselect list', () => {
  it('should be able to create by enter', () => {
    const onDeSelect = (v, o) => {
      expect(v).toBe('all');
      expect(o).toMatchObject({ label: '全部', value: 'all' });
    };

    const tree = mount(
      <Select multiple searchable options={optionsWithOutGroup} defaultValue={['all']} onDeselect={onDeSelect} />
    );
    act(() => {
      tree.simulate('click');
    });
    act(() => {
      tree.find('input').simulate('change', { target: { value: '全部' } });
    });
    act(() => {
      document.querySelector('.gio-select-dropdown').querySelectorAll('.gio-select-option')[0].click();
    });
    tree.unmount();
  });
});

describe('<Select /> deselect list', () => {
  it('should be able to create by enter', () => {
    const tree = mount(
      <Select searchable options={optionsWithOutGroup} defaultValue="all" onDeselect={onDeSelect} />
    );

    const onDeSelect = (v, o) => {
      expect(v).toBe('all');
      expect(o).toMatchObject({ label: '全部', value: 'all' });
      act(() => {
        tree.unmount();
      });
    };

    act(() => {
      tree.simulate('click');
    });
    act(() => {
      tree.find('input').simulate('change', { target: { value: '全部' } });
    });
    act(() => {
      document.querySelector('.gio-select-dropdown').querySelectorAll('.gio-select-option')[0].click();
    });
  });
});

describe('independent function should work as expected', () => {
  it('CutomOption should render free option with group', () => {
    expect(CustomOption('test', true)).toEqual({
      value: 'test',
      label: 'test',
      groupValue: 'select_custom_option_',
      groupLabel: '自由输入',
    });
  });

  it('label renderer should render highlighted item', () => {
    const option = {
      value: 'test',
      label: 'test',
      groupValue: 'select_custom_option_',
      groupLabel: '自由输入',
    };
    expect(defaultLabelRenderer('s')(option, true)).toBe('test');
    const tree = shallow(defaultLabelRenderer('s', 'select')(option));
    expect(tree.find('.select-search-highlight').text()).toBe('s');
  });
});
