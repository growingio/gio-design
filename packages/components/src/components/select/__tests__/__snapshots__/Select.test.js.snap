// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders <Select /> components 1`] = `
<div
  aria-controls="expandable"
  aria-disabled={false}
  aria-expanded={false}
  className="gio-select gio-select-middle gio-select-single gio-select-bordered"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  role="combobox"
  style={
    Object {
      "cursor": "pointer",
    }
  }
>
  <div
    className="gio-select-selector"
  >
    <div
      className="gio-select-values-wrapper"
    >
      <div
        className="gio-select-item"
      >
        <span
          className="gio-select-item-text"
        >
          全部
        </span>
      </div>
    </div>
  </div>
  <div
    className="gio-select-arrow"
  >
    <span
      className="gio-icon"
    >
      <svg
        className="gio-icon-svg"
        fill="currentColor"
        height="1rem"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 14 14"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <style />
        </defs>
        <g
          data-name="\\\\u56FE\\\\u5C42 2"
          id="down-filled_svg__\\\\u56FE\\\\u5C42_2"
        >
          <g
            data-name="\\\\u56FE\\\\u5C42 1"
            id="down-filled_svg__\\\\u56FE\\\\u5C42_1-2"
          >
            <path
              d="M7.39 9.51a.5.5 0 01-.78 0l-3-3.7a.51.51 0 01-.06-.53A.5.5 0 014 5h6a.5.5 0 01.45.28.51.51 0 01-.06.53z"
              fill="currentColor"
              fillRule="evenodd"
              id="down-filled_svg__down-2"
            />
          </g>
        </g>
      </svg>
    </span>
  </div>
</div>
`;

exports[`<Select /> renders <Select /> components 2`] = `
<div
  aria-controls="expandable"
  aria-disabled={false}
  aria-expanded={false}
  className="gio-select gio-select-middle gio-select-single gio-select-bordered"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  role="combobox"
  style={
    Object {
      "cursor": "pointer",
    }
  }
>
  <div
    className="gio-select-selector"
  >
    <div
      className="gio-select-values-wrapper"
    >
      <div
        className="gio-select-item"
      >
        <span
          className="gio-select-item-text"
        >
          全部
        </span>
      </div>
    </div>
  </div>
  <div
    className="gio-select-arrow"
  >
    <span
      className="gio-icon"
    >
      <svg
        className="gio-icon-svg"
        fill="currentColor"
        height="1rem"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 14 14"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <style />
        </defs>
        <g
          data-name="\\\\u56FE\\\\u5C42 2"
          id="down-filled_svg__\\\\u56FE\\\\u5C42_2"
        >
          <g
            data-name="\\\\u56FE\\\\u5C42 1"
            id="down-filled_svg__\\\\u56FE\\\\u5C42_1-2"
          >
            <path
              d="M7.39 9.51a.5.5 0 01-.78 0l-3-3.7a.51.51 0 01-.06-.53A.5.5 0 014 5h6a.5.5 0 01.45.28.51.51 0 01-.06.53z"
              fill="currentColor"
              fillRule="evenodd"
              id="down-filled_svg__down-2"
            />
          </g>
        </g>
      </svg>
    </span>
  </div>
</div>
`;

exports[`<Select multiple/> renders <Select multiple/> components 1`] = `
<div
  aria-controls="expandable"
  aria-disabled={false}
  aria-expanded={false}
  className="gio-select gio-select-middle gio-select-bordered"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  role="combobox"
  style={
    Object {
      "cursor": "pointer",
    }
  }
>
  <div
    className="gio-select-selector"
  >
    <div
      className="gio-select-values-wrapper"
    >
      <span
        aria-hidden="true"
        className="gio-tag gio-tag-size-middle gio-tag-type-normal gio-tag-closable gio-select-item"
      >
        <span
          className="gio-tag-label"
        >
          全部
        </span>
        <span
          className="gio-icon gio-tag-closable-icon"
          onClick={[Function]}
        >
          <svg
            className="gio-icon-svg"
            fill="currentColor"
            height="1rem"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 14 14"
            width="1rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <style />
            </defs>
            <g
              data-name="\\\\u56FE\\\\u5C42 2"
              id="close_svg__\\\\u56FE\\\\u5C42_2"
            >
              <g
                data-name="\\\\u56FE\\\\u5C42 1"
                id="close_svg__\\\\u56FE\\\\u5C42_1-2"
              >
                <path
                  d="M6.62 5.92a.54.54 0 00.76 0l3.87-3.87a.49.49 0 01.7.7L8.08 6.62a.54.54 0 000 .76L12 11.25a.49.49 0 01-.7.7L7.38 8.08a.54.54 0 00-.76 0L2.75 12a.49.49 0 01-.7-.7l3.87-3.92a.54.54 0 000-.76L2.05 2.75a.49.49 0 01.7-.7z"
                  fill="currentColor"
                  fillRule="evenodd"
                  id="close_svg__close"
                />
              </g>
            </g>
          </svg>
        </span>
      </span>
    </div>
  </div>
  <div
    className="gio-select-arrow"
  >
    <span
      className="gio-icon"
    >
      <svg
        className="gio-icon-svg"
        fill="currentColor"
        height="1rem"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 14 14"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <style />
        </defs>
        <g
          data-name="\\\\u56FE\\\\u5C42 2"
          id="down-filled_svg__\\\\u56FE\\\\u5C42_2"
        >
          <g
            data-name="\\\\u56FE\\\\u5C42 1"
            id="down-filled_svg__\\\\u56FE\\\\u5C42_1-2"
          >
            <path
              d="M7.39 9.51a.5.5 0 01-.78 0l-3-3.7a.51.51 0 01-.06-.53A.5.5 0 014 5h6a.5.5 0 01.45.28.51.51 0 01-.06.53z"
              fill="currentColor"
              fillRule="evenodd"
              id="down-filled_svg__down-2"
            />
          </g>
        </g>
      </svg>
    </span>
  </div>
</div>
`;

exports[`<Select multiple/> renders <Select multiple/> components 2`] = `
<div
  aria-controls="expandable"
  aria-disabled={false}
  aria-expanded={false}
  className="gio-select gio-select-middle gio-select-bordered"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  role="combobox"
  style={
    Object {
      "cursor": "pointer",
    }
  }
>
  <div
    className="gio-select-selector"
  >
    <div
      className="gio-select-values-wrapper"
    >
      <span
        aria-hidden="true"
        className="gio-tag gio-tag-size-middle gio-tag-type-normal gio-tag-closable gio-select-item"
      >
        <span
          className="gio-tag-label"
        >
          全部
        </span>
        <span
          className="gio-icon gio-tag-closable-icon"
          onClick={[Function]}
        >
          <svg
            className="gio-icon-svg"
            fill="currentColor"
            height="1rem"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 14 14"
            width="1rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <style />
            </defs>
            <g
              data-name="\\\\u56FE\\\\u5C42 2"
              id="close_svg__\\\\u56FE\\\\u5C42_2"
            >
              <g
                data-name="\\\\u56FE\\\\u5C42 1"
                id="close_svg__\\\\u56FE\\\\u5C42_1-2"
              >
                <path
                  d="M6.62 5.92a.54.54 0 00.76 0l3.87-3.87a.49.49 0 01.7.7L8.08 6.62a.54.54 0 000 .76L12 11.25a.49.49 0 01-.7.7L7.38 8.08a.54.54 0 00-.76 0L2.75 12a.49.49 0 01-.7-.7l3.87-3.92a.54.54 0 000-.76L2.05 2.75a.49.49 0 01.7-.7z"
                  fill="currentColor"
                  fillRule="evenodd"
                  id="close_svg__close"
                />
              </g>
            </g>
          </svg>
        </span>
      </span>
    </div>
  </div>
  <div
    className="gio-select-arrow"
  >
    <span
      className="gio-icon"
    >
      <svg
        className="gio-icon-svg"
        fill="currentColor"
        height="1rem"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 14 14"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <style />
        </defs>
        <g
          data-name="\\\\u56FE\\\\u5C42 2"
          id="down-filled_svg__\\\\u56FE\\\\u5C42_2"
        >
          <g
            data-name="\\\\u56FE\\\\u5C42 1"
            id="down-filled_svg__\\\\u56FE\\\\u5C42_1-2"
          >
            <path
              d="M7.39 9.51a.5.5 0 01-.78 0l-3-3.7a.51.51 0 01-.06-.53A.5.5 0 014 5h6a.5.5 0 01.45.28.51.51 0 01-.06.53z"
              fill="currentColor"
              fillRule="evenodd"
              id="down-filled_svg__down-2"
            />
          </g>
        </g>
      </svg>
    </span>
  </div>
</div>
`;
